lib/rules/no-loop-func.js

```diff
@@ -186,13 +199,58 @@ module.exports = {
             }
 
             const references = sourceCode.getScope(node).through;
-            const unsafeRefs = references.filter(r => r.resolved && !isSafe(loopNode, r)).map(r => r.identifier.name);
+            const unsafeRefs = references.filter(r => r.resolved && !isSafe(loopNode, r));
+            const unsafeRefsName = unsafeRefs.map(r => r.identifier.name);
+
+            // Check if the function is not asynchronous or a generator function
+            if (!(node.async || node.generator)) {
+                const isFunctionExpression = node.type === "FunctionExpression";
+
+                if (isIIFE(node)) {
+
+                    // Check if the function is referenced elsewhere in the code
+                    const isFunctionReferenced = isFunctionExpression && node.id ? references.some(r => r.identifier.name === node.id.name) : false;
+
+                    // Check if there are unsafe references used within non-immediately-invoked nested functions
+                    const unsafeRefsInNonIIFE = unsafeRefs.filter(r => {
+                        let refScope = r.from.variableScope;
+
+                        if (node === refScope.block) {
+                            return !isIIFE(refScope.block);
+                        }
+
+                        while (node !== refScope.block) {
+                            if (!isIIFE(refScope.block)) {
+                                return true;
+                            }
+                            refScope = refScope.upper.variableScope;
+                        }
+
+                        return false;
+                    });
+
+                    const unsafeRefsInNonIIFEName = unsafeRefsInNonIIFE.map(r => r.identifier.name);
+
+                    if (!isFunctionReferenced && unsafeRefsInNonIIFEName.length === 0) {
+                        return;
+                    }
+
+                    if (unsafeRefsInNonIIFEName.length > 0) {
+                        context.report({
+                            node,
+                            messageId: "unsafeRefs",
+                            data: { varNames: `'${unsafeRefsInNonIIFEName.join("', '")}'` }
+                        });
+                        return;
+                    }
+                }
+            }
```

Comment:
Apologies for the delayed respose. Do you mean we can remove this whole logic?

Reply:
Yes, I think we can remove all code that handles inner functions of skipped ones as a special case. But we should also update the code that checks if a function is already in a function that is in a loop (to avoid duplicate reports) to also skip the same functions.